### PR TITLE
Tests: Add tests that will generate more coverage

### LINKED ISSUE #88: Improve test coverage
## Description
Currently, the test coverage is around 88% which is okay but we can improve it. This issue focuses on adding tests covering missing branches.

## People to notify
@manrajgrover

### PR DESCRIPTION
**Pull request by Lilja**
## Description of new feature, or changes
Adds more tests and coverage

## Checklist

- [x] Your branch is up-to-date with the base branch
- [x] You've included at least one test if this is a new feature
- [x] All tests are passing

## Related Issues and Discussions
Fixes #88 

## People to notify
@manrajgrover

### CODE REVIEW

**Comment by Lilja** [comment]
```
.................................................
Name                    Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------------
halo/__init__.py            6      0      0      0   100%
halo/_utils.py             46      9     10      3    79%   9-10, 44-45, 51, 53, 57-58, 130, 50->51, 52->53, 129->130
halo/halo.py              181      6     46      4    96%   79, 82-85, 394, 423, 81->82, 393->394, 414->418, 422->423
halo/halo_notebook.py      57      2     12      2    94%   55, 84, 54->55, 83->84
-------------------------------------------------------------------
TOTAL                     290     17     68      9    93%
----------------------------------------------------------------------
```

File: tests/test_halo.py [hunk 1]
```diff
@@ -71,6 +79,29 @@ def test_basic_spinner(self):
...
+        instance.spinner = default_spinner_value
+        self.assertEqual(default_spinner, instance.spinner)
+
+        instance.spinner = "This_spinner_do_not_exist"
+        self.assertEqual(default_spinner, instance.spinner)
+
+        instance.spinner = -123
+        self.assertEqual(default_spinner, instance.spinner)
+
+    def test_windows_users_get_their_default_spinner(self):
```
Comment (manrajgrover) [inline tests/test_halo.py]: We shouldn't add OS specific tests. We run the code on Travis(Linux) and Appveyor(Windows) for testing in both operating systems.

File: tox.ini [hunk 2]
```diff
@@ -9,6 +9,21 @@ envlist =
 skip_missing_interpreters =
     True
 
+[testenv:py27]
```
Comment (manrajgrover) [inline tox.ini]: This will get removed because OS specific tests are not needed

File: halo/halo.py [hunk 3]
```diff
@@ -14,7 +14,8 @@
 from spinners.spinners import Spinners
 from log_symbols.symbols import LogSymbols
 
-from halo._utils import is_supported, colored_frame, is_text_type, decode_utf_8_text, get_terminal_columns, \
+from halo import _utils
```
Comment (manrajgrover) [inline halo/halo.py]: Use `halo._utils` to import all utils or revert this change. I don't see the point of importing this separately.

File: halo/halo.py [hunk 4]
```diff
@@ -82,7 +83,7 @@ def clean_up():
 
             ip = get_ipython()
             ip.events.register('post_run_cell', clean_up)
-        elif environment == 'terminal':
+        else:  # default terminal
```
Comment (manrajgrover) [inline halo/halo.py]: Thanks for fixing this

File: tests/test_halo.py [hunk 5]
```diff
@@ -5,21 +5,28 @@
 import os
 import re
 import sys
+import platform
 import time
 import unittest
+try:
```
Comment (manrajgrover) [inline tests/test_halo.py]: I don't think we need this

File: tests/test_halo.py [hunk 6]
```diff
@@ -5,21 +5,28 @@
...
+    from mock import Mock
+
 
 from spinners.spinners import Spinners
 
 from tests._utils import strip_ansi, encode_utf_8_text, decode_utf_8_text
 from halo import Halo
-from halo._utils import is_supported, get_terminal_columns
+from halo._utils import get_terminal_columns
+from halo import _utils
```
Comment (manrajgrover) [inline tests/test_halo.py]: Same here, we can use `halo._utils`

File: tests/test_halo.py [hunk 7]
```diff
@@ -5,21 +5,28 @@
 import os
 import re
 import sys
+import platform
 import time
 import unittest
+try:
+    from unittest.mock import Mock
+except: # Py 2.7
+    from mock import Mock
```
Comment (manrajgrover) [inline tests/test_halo.py]: This will be reverted. See the comment below.

**Comment by manrajgrover** [comment]
@Lilja Looks like we are not testing utilities well. We can improve on that if you would like to take it up in this PR

**Comment by Lilja** [comment]
@manrajgrover Thanks for your feedback. Do you want me to create multiple commits or stay strict to just one?

**Comment by manrajgrover** [comment]
@Lilja > Thanks for your feedback. Do you want me to create multiple commits or stay strict to just one?

Please add more commits to this PR only. Don't worry about that.

> My analysis of _utils.py is that 90% of the untested stuff is related to ipython or jython. I've have no clue what they are.

It would be ipython. Testing it would be similar to testing it in terminal except for text width (if I remember correctly)

> Also, some of the coverage will not be reported on windows whereas on linux it will be reported. So the coverage report needs to be a merge of these two. You might already know this but I'd figure that I'd just tell you anyway!

Yup, I'm aware of that :) Thanks for remind though

**Comment by Lilja** [comment]
@manrajgrover Allright. Maybe Ipython tests could be another PR?

**Comment by manrajgrover** [comment]
@Lilja Yup, you can raise a separate PR for it
